### ISSUE TITLE
NSFetchRequest not working after Seam3 setup

### ISSUE THREAD

**Issue by DJ-Glock** [paulw11/Seam3#37]
Hello Paul.
I'm still having an issue with Seam3 integration to my project :(
I have setup per your manual and example and noticed that after that my fetch requests stop working. All the time I'm getting no data.

Here is an example of my request:
```
    static let tableSessionRequest: NSFetchRequest<TableSessionTable> = TableSessionTable.fetchRequest()

    class func getCurrentTableSession (table: TablesTable) -> TableSessionTable? {
        tableSessionRequest.predicate = NSPredicate(format: "table = %@ and closeTime = %@", table, NSNull() as CVarArg)
        let matchedTableSession = try? context.fetch(tableSessionRequest)
        if matchedTableSession?.count == 1 {
            return matchedTableSession?[0]
        }
        else {
            return nil
        }
    }
```
In the same time inserts work with no issues. Synchronization works too - at least I can see that new records appear in CloudKit dashboard.
Example of function that performs insert into same table:
```
    class func createTableSession (table: TablesTable) -> TableSessionTable {
        if #available(iOS 10.0, *) {
            let newTableSession = TableSessionTable(context: context)
            newTableSession.openTime = Date() as NSDate
            newTableSession.closeTime = nil
            table.addToTableSession(newTableSession)
            try? context.save()
            return newTableSession
        } else {
            let newTableSession = TableSessionTable(entity: NSEntityDescription.entity(forEntityName: "TableSessionTable", in: context)!, insertInto: context)
            newTableSession.openTime = Date() as NSDate
            newTableSession.closeTime = nil
            table.addToTableSession(newTableSession)
            try? context.save()
            return newTableSession
        }
    }
```
And here is my AppDelegate. For now Seam3 is setup for iOS 10+.
```
import UIKit
import CoreData
import CloudKit
import Seam3

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    var smStore: SMStore?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        
        if #available(iOS 10.0, *) {
            let container = AppDelegate.persistentContainer
            self.smStore = container.persistentStoreCoordinator.persistentStores.first as? SMStore
        } else {
            // Fallback on earlier versions
        }
        
        application.registerForRemoteNotifications()
        
        // Uncomment the following block to use `clientTellsWhichWins` conflict resolution
        
        self.smStore?.recordConflictResolutionBlock =  ({(serverRecord, clientRecord, ancestorRecord) -> CKRecord in
            
            if let serverInt = serverRecord["intAttribute"] as? NSNumber,
                let clientInt = clientRecord["intAttribute"] as? NSNumber {
                
                if clientInt.intValue > serverInt.intValue {
                    serverRecord["intAttribute"] = clientInt
                }
                
            }
            
            
            return serverRecord
            
        } )
        
        
        self.validateCloudKitAndSync() {}
        
        return true
    }
    
    func validateCloudKitAndSync(_ completion:@escaping (() -> Void)) {
        self.smStore?.verifyCloudKitConnectionAndUser() { (status, user, error) in
            guard status == .available, error == nil else {
                NSLog("Unable to verify CloudKit Connection \(error!)")
                return
            }
            guard let currentUser = user else {
                NSLog("No current CloudKit user")
                return
            }
            let previousUser = UserDefaults.standard.string(forKey: "CloudKitUser")
            if  previousUser != currentUser {
                do {
                    print("New user")
                    try self.smStore?.resetBackingStore()
                } catch {
                    NSLog("Error resetting backing store - \(error.localizedDescription)")
                    return
                }
            }
            UserDefaults.standard.set(currentUser, forKey:"CloudKitUser")
            self.smStore?.triggerSync(complete: true)
            completion()
        }
    }
    
    
    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }
    
    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
        // Saves changes in the application's managed object context before the application terminates.
        if #available(iOS 10.0, *) {
            self.saveContext()
        } else {
            self.saveOldStyleContext()
        }
    }
    
    //Notifications
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any]) {
        self.smStore?.handlePush(userInfo: userInfo)
    }
    
    
    // MARK: - Core Data stack
    @available(iOS 10.0, *)
    static var persistentContainer: NSPersistentContainer = {
        SMStore.registerStoreClass()
        let container = NSPersistentContainer(name: "CafeManager")
        let urls = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        if let applicationDocumentsDirectory = urls.last {
            let url = applicationDocumentsDirectory.appendingPathComponent("SingleViewCoreData.sqlite")
                print(url)
            let storeDescription = NSPersistentStoreDescription(url: url)
            storeDescription.type = SMStore.type
            container.persistentStoreDescriptions=[storeDescription]
            container.loadPersistentStores(completionHandler: { (storeDescription, error) in
                if let error = error as NSError? {
                    fatalError("Unresolved error \(error), \(error.userInfo)")
                }
            })
            print(container.persistentStoreDescriptions)
            return container
        }
        
        fatalError("Unable to access documents directory")
        
    }()
    
    static var viewContext: NSManagedObjectContext {
        if #available(iOS 10.0, *) {
            return persistentContainer.viewContext
        } else {
            return AppDelegate.managedObjectContext
        }
    }
    
    // MARK: - Core Data stack for iOS 8+
    static var managedObjectContext: NSManagedObjectContext = {
        
        var applicationDocumentsDirectory: URL = {
            // The directory the application uses to store the Core Data store file. This code uses a directory named "com.cadiridris.coreDataTemplate" in the application's documents Application Support directory.
            let urls = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
            return urls[urls.count-1]
        }()
        
        var managedObjectModel: NSManagedObjectModel = {
            // The managed object model for the application. This property is not optional. It is a fatal error for the application not to be able to find and load its model.
            let modelURL = Bundle.main.url(forResource: "CafeManager", withExtension: "momd")!
            return NSManagedObjectModel(contentsOf: modelURL)!
        }()
        
        var persistentStoreCoordinator: NSPersistentStoreCoordinator = {
            // The persistent store coordinator for the application. This implementation creates and returns a coordinator, having added the store for the application to it. This property is optional since there are legitimate error conditions that could cause the creation of the store to fail.
            // Create the coordinator and store
            let coordinator = NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel)
            //Name of sqlite file was not changed initially. Don't change it to not affect all users of iOS 9. :(
            let url = applicationDocumentsDirectory.appendingPathComponent("SingleViewCoreData.sqlite")
            var failureReason = "There was an error creating or loading the application's saved data."
            do {
                try coordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: url, options: nil)
            } catch {
                // Report any error we got.
                var dict = [String: AnyObject]()
                dict[NSLocalizedDescriptionKey] = "Failed to initialize the application's saved data" as AnyObject?
                dict[NSLocalizedFailureReasonErrorKey] = failureReason as AnyObject?
                
                dict[NSUnderlyingErrorKey] = error as NSError
                let wrappedError = NSError(domain: "iGlock.CafeManager.com", code: 9999, userInfo: dict)
                // Replace this with code to handle the error appropriately.
                // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
                NSLog("Unresolved error \(wrappedError), \(wrappedError.userInfo)")
                abort()
            }
            
            return coordinator
        }()
        
        // Returns the managed object context for the application (which is already bound to the persistent store coordinator for the application.) This property is optional since there are legitimate error conditions that could cause the creation of the context to fail.
        let coordinator = persistentStoreCoordinator
        var managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
        managedObjectContext.persistentStoreCoordinator = coordinator
        return managedObjectContext
    }()
    
    // MARK: - Core Data Saving support for iOS 10
    @available(iOS 10.0, *)
    func saveContext () {
        let context = AppDelegate.persistentContainer.viewContext
        if context.hasChanges {
            do {
                try context.save()
            } catch {
                // Replace this implementation with code to handle the error appropriately.
                // fatalError() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
                let nserror = error as NSError
                fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
            }
        }
    }
    
    // MARK: - Core Data Saving support for iOS 8+
    func saveOldStyleContext () {
        if AppDelegate.managedObjectContext.hasChanges {
            do {
                try AppDelegate.managedObjectContext.save()
            } catch {
                // Replace this implementation with code to handle the error appropriately.
                // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
                let nserror = error as NSError
                NSLog("Unresolved error \(nserror), \(nserror.userInfo)")
                abort()
            }
        }
    }
}
```
If I rollback AppDelegate to my version, it works with no issues. But with Seam3, it looks like all fetch requests are broken. I thought that cause an be in url of sqllite file, but not sure. I will try to debug it step by step carefully, but any advice will be highly appreciated.

P.S. I think you can check my AppDelegate to fix your help page a bit. For example, from your page:
var smStore: SMStore - it should be optional I believe, as in your AppDelegate example.

**Comment by DJ-Glock** [comment]
Update: It looks like fetch requests are not working for Predicates with reference to another table. If I use no predicate or predicate without reference to another table - it works fine.

**Comment by DJ-Glock** [comment]
Hey
Okay. Finally I had some time to dive into Seam3 code. I have added some prints and looks like I have found why it is not working for my fetch requests with links to another tables. Let me share my findings.

I have put prints here:
```
    override open func execute(_ request: NSPersistentStoreRequest, with context: NSManagedObjectContext?) throws -> Any {
        print("request")
        print(request)
        print("end of request")
        if request.requestType == NSPersistentStoreRequestType.fetchRequestType {
```

So when I run my app and perform some actions, it gave me this output:

> file:///Users/dj-glock/Library/Developer/CoreSimulator/Devices/AAD53D28-3CD3-4BA9-A611-13F6489CAD3D/data/Containers/Data/Application/0B8F7EB5-426C-4E83-B657-C174E1186571/Documents/CafeManagerSeam3.sqlite
> Sync Started
> No more records coming
> Sync Performed
> Sync performed successfully
> request
> <NSFetchRequest: 0x6000000cc780> (entity: TablesTable; predicate: ((null)); sortDescriptors: ((
>     "(tableName, ascending, localizedCaseInsensitiveCompare:)"
> )); type: NSManagedObjectResultType; )
> end of request
> request
> <NSFetchRequest: 0x6080000d0c30> (entity: TableSessionTable; predicate: (table == <CafeManager.TablesTable: 0x6080002998c0> (entity: TablesTable; id: 0x608000236a00 <x-coredata://EB4E0B6C-CE74-4117-ABB8-BCAB81F05799-18890-0000332D66F92AA7/TablesTable/pE59EE99F-CB80-48C1-807A-65DF699BDE3E> ; data: <fault>) AND closeTime == <null>); sortDescriptors: ((null)); type: NSManagedObjectResultType; )
> end of request

The first fetch request executed successfully and I've got some records. But the second returned no rows.
I have checked my sqlite database and can see that it looks like some function added "p" character to record id. And this caused no matches I think. 

![logwithp](https://user-images.githubusercontent.com/26511191/30329379-c76e38ea-97da-11e7-92d8-bc0d355d74c1.png)
![databasewithseam](https://user-images.githubusercontent.com/26511191/30329918-6c3342e8-97dc-11e7-8aad-bf42f71f97cd.png)

I have compared fetch request that is generated without Seam, it also has "p" character, but as I can see there is Z_PK value after "p".

Example:

> <NSFetchRequest: 0x6000000d30f0> (entity: TableSessionTable; predicate: (table == <CafeManager.TablesTable: 0x600000893ec0> (entity: TablesTable; id: 0xd000000000080000 <x-coredata://87826087-8DE4-45B7-A70E-FC2163B8EB0F/TablesTable/p2> ; data: <fault>) AND closeTime == <null>); sortDescriptors: ((null)); type: NSManagedObjectResultType; )
> <NSFetchRequest: 0x6000000d30f0> (entity: TableSessionTable; predicate: (table == <CafeManager.TablesTable: 0x600000893f10> (entity: TablesTable; id: 0xd0000000000c0000 <x-coredata://87826087-8DE4-45B7-A70E-FC2163B8EB0F/TablesTable/p3> ; data: <fault>) AND closeTime == <null>); sortDescriptors: ((null)); type: NSManagedObjectResultType; )
> <NSFetchRequest: 0x6000000d30f0> (entity: TableSessionTable; predicate: (table == <CafeManager.TablesTable: 0x60800028d660> (entity: TablesTable; id: 0xd000000000040000 <x-coredata://87826087-8DE4-45B7-A70E-FC2163B8EB0F/TablesTable/p1> ; data: {
>     tableCapacity = 6;
>     tableDescription = nil;
>     tableName = "Test 1";
>     tableSession =     (
>         "0xd000000000040002 <x-coredata://87826087-8DE4-45B7-A70E-FC2163B8EB0F/TableSessionTable/p1>"
>     );
> }) AND closeTime == <null>); sortDescriptors: ((null)); type: NSManagedObjectResultType; )

![databasenoseam](https://user-images.githubusercontent.com/26511191/30329428-e41068c4-97da-11e7-997f-994e0fdbdda2.png)


I suppose that predicate should be like this:
`let predicate = NSPredicate(format: "%K == %@", SMStore.SMLocalStoreRecordIDAttributeName,recordID)`

like in this function:
`    override open func newValuesForObject(with objectID: NSManagedObjectID, with context: NSManagedObjectContext) throws -> NSIncrementalStoreNode {`

I think it should be fixed in Seam3 or I will have to rewrite all my fetch requests. Please correct me if I am wrong and advise.

**Comment by paulw11** [comment]
I am not sure what you mean by "references to other tables" - Are you talking about reference type attributes?  With Core Data you don't refer to tables; it is an object persistence system that may (or may not) use sqlLite as a store.

Where/how is the  predicate changed needed?  In Seam3 code or your code?

**Comment by DJ-Glock** [comment]
Hi
Sorry for confusion. I was talking about Core Data with sqlite database, when you have two tables referenced to each other. So i have a function that extracts record from table that has record in another table linked to it’s id. Simple inner join. It works with no issues without seam, but not working with seam.
Unfortunately I cannot share u more details, my mac failed yesterday, I’m on vacation and will need more time to provide you more detailed explanation.
Apologies.

**Comment by paulw11** [comment]
You should never access the underlying sqlite tables when using CoreData.  They should be treated as opaque data stores.  It happens that Seam3 uses a local CoreData store backed by sqllite but that should be treated as private.

If you need to manage relationships between objects in Core Data then you should use references in Core Data

**Comment by DJ-Glock** [comment]
I do!
So let me explain one more time. I solved issues with my Mac so now I can proceed.

I have database model with several tables with relationships. Here is an example of couple of them:
<img width="385" alt="2017-09-21 9 45 32" src="https://user-images.githubusercontent.com/26511191/30682263-a698a8a2-9eb1-11e7-8ff8-966831f6aa79.png">

To get TableSession object, that is referenced to specific Tables object, I am using this function:
```
    class func getCurrentTableSession (table: TablesTable) -> TableSessionTable? {
        tableSessionRequest.predicate = NSPredicate(format: "table = %@ and closeTime = %@", table, NSNull() as CVarArg)
        let matchedTableSession = try? context.fetch(tableSessionRequest)
        if matchedTableSession?.count == 1 {
            return matchedTableSession?[0]
        }
        else {
            return nil
        }
    }
```
And it works fine, without seam.
When I use Seam, it just returns nothing (because of some issue with recordID)

Am I wrong here using NSPredicate with reference?

**Comment by paulw11** [comment]
Core Data will follow references for you; `table.tableSession` will be a set of `TableSessionTable` entities.  You can then just iterate this set filtering by `openTime` or `closeTime` if you want to.  

Don't think of Core Data as a database.  It is an object persistence system. References are not foreign keys.  They are object references,  just like `view.subViews` is an array of references to `UIView` objects.  Core Data will fault the object into memory when it needs to.

**Comment by DJ-Glock** [comment]
Hello Paul.
Understood. Looks like it's my very big problem. I have been worked with Oracle Database for 6 years and not very good in object oriented programming. So I tried to work with CoreData like with database.

Thanks. I will spend many happy hours with refactoring my code :)

**Comment by tifroz** [comment]
Seeing the same problem identified by @DJ-Glock so I used the Example app to write a simple test, (the test code was inserted in `MasterViewController.loadData()`):

```
func loadData() {
      // ...stripped some code here for clarity sake
        do {
            // ...stripped more code
            if let events = try self.managedObjectContext?.fetch(fetchRequest){
                self.events = events
              
              // TEST BEGINS HERE
              if let event = events.first, let moc = managedObjectContext {
                let entityDescription = NSEntityDescription.entity(forEntityName: "Event", in: moc)
                let request = NSFetchRequest<NSFetchRequestResult>()
                request.entity = entityDescription
                let predicate = NSPredicate(format: "SELF == %@", event)
                request.predicate = predicate
                let matches = try! moc.fetch(request)
                
                NSLog("We should get exactly one match. Match count = \(matches.count)")
              }
              // END OF TEST
            }
            self.tableView.reloadData()
        } catch {}
    }
```

Here is the output, after hitting the (+) button to add an event:
```
New user
Registered for remote notifications
Created device id D159B882-C5D9-4E6A-B84D-8E1F4B63B401
We should get exactly one match. Match count = 0
```

The predicate tests for the existence of an `event` that we know for sure is present because we just obtained this `event` from an earlier fetch request. However the predicated fetch request fails to find it.

@paulw11  Happy to create a pull request if you can point me in the right direction?

**Comment by paulw11** [comment]
The problem is that the predicate is passing through an identifier for the event in the Seam virtual store, not the identifier for the event in the Seam backing store, and NSPredicate does not offer a convenient method to determine the values that need modification.

Generally if you have the inverse relationship set up in Core Data you should not need to perform a query like this, since the objects you need can be obtained simply by retrieving the relationship value you already have.

I can see that it might be useful to combine this predicate with some further filtering, but again you can potentially do this in memory once you have retrieved the full list of related objects.

The only other solution is to query the internal key which is a bit "hacky".  

This works:
```
let appDelegate = UIApplication.shared.delegate as! AppDelegate
                        
                        
  let recordID = appDelegate.smStore!.referenceObject(for: event.objectID)
                        
  let predicateObjectRecordIDKey = "objectRecordID"
  let predicate: NSPredicate = NSPredicate(format: "%K == $objectRecordID", "sm_LocalStore_RecordID")
                        
  request.predicate =  predicate.withSubstitutionVariables([predicateObjectRecordIDKey:recordID])
  let matches = try! moc.fetch(request)
                        
   NSLog("We should get exactly one match. Match count = \(matches.count)")
```

**Comment by paulw11** [comment]
Here is a new approach; it is still a bit hacky, but it is the best I can think of.

There is a new method on `SMStore` `predicate(for relationship:String, referencing object:NSManagedObject)`  that will return an `NSPredicate` that will match where the specified relationship name refers to the specified object.

The is another new method predicate(forToMany relationship:String, referencing object:NSManagedObject)`  that will return an `NSPredicate` that will match where the specified to-many relationship name contains a reference to the specified object.

e.g.

``` 
if  let moc = managedObjectContext {
     let entityDescription = NSEntityDescription.entity(forEntityName: "Event", in: moc)
                
    let request = NSFetchRequest<Event>()
    request.entity = entityDescription
                
    request.predicate = appDelegate.smStore!.predicate(for: "creatingDevice", referencing: self.device!)
                
    let matches = try! moc.fetch(request)
                
    for event in matches {
        print(event.timestamp ?? "No timestamp")
    }
}
```

and
```
let entityDescription = NSEntityDescription.entity(forEntityName: "Device", in: moc)
                
let request = NSFetchRequest<Device>()
request.entity = entityDescription
                
request.predicate = appDelegate.smStore!.predicate(forToMany: "events", referencing: self.detailItem!)

let matches = try! moc.fetch(request)
                
for event in matches {
    print(event.deviceID ?? "No device")
}
```

**Comment by paulw11** [comment]
Here is a new approach; it is still a bit hacky, but it is the best I can think of.

There is a new function on `SMStore` `predicate(for relationship: String, referencing object: NSManagedObject) -> NSPredicate`  that will return an `NSPredicate` that will match where  the specified relationship matches the provided object.

A second new function `predicate(forToMany relationship: String, referencing object: NSManagedObject) -> NSPredicate` returns a predicate that will match where the specified to-many relationship *contains* a reference to the provided object.

You can combine the returned predicates with other conditions using an `NSCompoundPredicate`

**Comment by paulw11** [comment]
These new methods are now in the repo; if you get a chance can you update from the repo instead of the released POD and see if it works for you.

**Comment by tifroz** [comment]
Will do

**Comment by tifroz** [comment]
I have only been performing some pretty basic tests and so far so good. I will share an update in 2-3 days when data & my code are in better shape and I am in a position to perform more meaningful testing

**Comment by Jerland2** [comment]
I am not seeing this method on SMStore.  How can I access the smStore.predicate.. methods?

**Comment by DJ-Glock** [comment]
@Jerland2 Hi
For example you can define var smStore in AppDelegate 
`    var smStore: SMStore?
`and use like this:
`let referencePredicate = appDelegate.smStore?.predicate(for: "table", referencing: currentTable!)`

**Comment by Jerland2** [comment]
@DJ-Glock I thats what I was expecting however when I access SMStore there is no `predicate(for: "table", referencing: currentTable!)`

to confirm in my pod file I am using "pod Seam3", do I have to point to a different branch? I am also importing Seam3 in the file I want to use the predicate method in

~~Edit: Error specifically is:
`Value of type 'SMStore?' has no member 'predicate'`~~

~~Edit2: As I suspected those methods are missing from my SMStore.Swift file.  Any thoughts on how I pull the correct branch? or which branch would include those methods?  Again my pod file line is referenced above.  Thanks!~~

~~Edit3: Updated my pod file and now have the latest version of the repo.  However this method crashes my app.  Have you had any success using these methods with a Fetched Results Controller?~~

Edit4: It Works!! However now when adding an object to my Fetched Results Controller it does not auto populate in my tableview, this must be a byproduct of something I just changed as this did not occur before.  It seems that the predicate is interfering with my FRC's delegate because I can print the FRC.delegate and get the correct view controller, however none of the methods trigger.  The predicate workaround seems to be the culprit after reading this SOF: https://stackoverflow.com/questions/17151160/nsfetchedresultscontroller-delegate-not-called-for-api-listener

Unsure whether to reopen this issue, or to open a new issue, predicates breaking Fetched result controllers seems like a major problem as they are the preferred way to populate tableViews with Core Data?
